feat: typeORM을 이용한 모델 생성

Do-ho opened the pull request:
## 📕 제목

typeORM을 이용한 모델 생성
관련 이슈 #25 

## 📗 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] 여러 엔티티를 받기 위한 파일 수정
- [x] User에 대한 모델 설정
- [x] Issue에 대한 모델 설정
- [x] Milestone에 대한 모델 설정
- [x] Label에 대한 모델 설정
- [x] Comment에 대한 모델 설정
- [x] IssueAssignee에 대한 모델 설정
- [x] IssueLabel에 대한 모델 설정
- [x] class-validator를 활용한 모델 설정


## 📘 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- OneToMany, ManyToOne 관계는 모두 양측에 선언을 해주었습니다.
- ManyToMany 관계는 따로 테이블을 두어 OneToMany, ManyToOne로 바꿔주었습니다.
- 기본적으로 [typeORM 공식 문서](https://typeorm.io/#/)를 보고 작성하였습니다. 
- 같이 해주신 assignee분들은 다시 한 번 검토 부탁드립니다!
- enum, cascade, eager loading 관련해서는 계속적인 issue로 다뤄졌으면 합니다.

Do-ho committed:
chore: 여러 엔티티를 받기 위한 파일 수정

- entities에 User 엔티티를 받고 있던 코드를 파일 경로로 변경함으로 여러 모델을 정의할 수 있도록 설정함

Do-ho committed:
feat: user 모델 정의

- DB 명세에 따른 모델 정의
- 관계에 대한 정의를 해주기 위해 OneToMany를 사용

Do-ho committed:
feat: label 모델 정의

- DB 명세에 따른 모델 정의
- 관계에 대한 정의를 해주기 위해 OneToMany를 사용

Do-ho committed:
feat: milestone 모델 정의

- DB 명세에 따른 모델 정의
- 관계에 대한 정의를 해주기 위해 OneToMany를 사용

Do-ho committed:
feat: comment 모델 정의

- DB 명세에 따른 모델 정의
- 관계에 대한 정의를 해주기 위해 ManyToOne를 사용

Do-ho committed:
feat: issue 모델 정의

- DB 명세에 따른 모델 정의
- 관계에 대한 정의를 해주기 위해 ManyToOne, OneToMany를 사용

Do-ho committed:
feat: label-issue 관계 모델 정의

- DB 명세에 따른 관계 모델 정의
- 관계에 대한 정의를 해주기 위해 ManyToOne를 사용

Do-ho committed:
feat: user-issue 관계 모델 정의

- DB 명세에 따른 관계 모델 정의
- 관계에 대한 정의를 해주기 위해 ManyToOne를 사용

wooojini reviewed (approved):
고생하셨습니다! 해당 부분을 토대로 테스트 코드를 작성해보겠습니다! :-)

jinhyukoo commented on backend/src/model/issue.js:
```diff
@@ -0,0 +1,59 @@
+    @Column({ name: "state", type: "varchar", default: STATE.OPEN})
+    @IsString()
+    state;
+
```
이 부분은 enum의 잔재인걸까요...!? 추후 sqlite3가 아닌 mysql로 갈아탈거라면 상관 없겠지만 아예 varchar로 갈거라면 수정을 해도 괜찮겠다라는 생각도 드네요! 근데 아마 enum으로 갈아타겠죠.. ㅎㅎ?

Do-ho replied:
default는 제가 따로 추가한 부분인데 일전에 리뷰를 받았을 때 ```default: "open"``` 과 같이 하드코딩된 정보는
```js
const STATE = {
    OPEN: "open",
    CLOSED: "closed"
}
```
과 같이 관리해주는 게 좋다고 해서 변수화 시켰습니다! 혹시 문제가 될까요?

jinhyukoo replied:
CLOSED가 쓰이지 않고 있는 것 같아서 언급해봤습니다!

Do-ho replied:
나중에 API 만들 때 closed 된 상황에 따른 변수가 필요할 수 있으니 따로 common 폴더의 변수로 빼도 좋을 것 같네요!

jinhyukoo commented on backend/src/model/label.js:
```diff
@@ -0,0 +1,39 @@
+    @Column({ name: "name", type: "varchar", unique: true })
+    @IsString()
+    name;
+
```
혹시나 하고 깃허브에서 실험해봤는데 한글로도 라벨이 만들어지네요! charset: "utf-8"을 추가해주면 좋을 것 같습니다!

Do-ho replied:
리뷰 반영했습니다! 감사합니다 👍

jinhyukoo reviewed (approved):
지금 바로 수정해야 하는 큰 문제들은 아닐뿐더러 빠르게 작업을 해야하니 일단은 Approve를 하겠습니당! 감사합니다!!

jinhyukoo commented:
Cascade에 관한 것도 한 번 더 이야기를 해봐야겠군요..!

Do-ho committed:
refactor: model prettier 미반영 문제 해결

- 리뷰에 따른 utf-8 추가
- prettier 반영

youngxpepp commented on backend/src/common/config/database/connection-option-generator.js:
```diff
@@ -11,7 +10,7 @@ class ConnectionOptionGenerator {
         const connectionOption = {
             type: this.databaseEnv.getDatabaseType(),
-            entities: [User],
+            entities: ["src/model/*.js"],
```
npm run build -> npm run start:prod
위와 같이 실행해도 잘 동작하나요?

jinhyukoo replied:
경로를 "dist/model/*.js"로 바꿔주면 잘 동작합니다! 이 부분에 대해서 개발환경과 배포환경에 따라 경로를 다르게 지정해줄 수 있으면 좋을 것 같아요!

youngxpepp commented on backend/src/model/issue.js:
```diff
@@ -0,0 +1,59 @@
+import { Milestone } from "./milestone";
+import { LabelToIssue } from "./label-to-issue";
+
+const STATE = {
+    OPEN: "open",
+    CLOSED: "closed"
+};
```
비지니스 로직에서 사용될  확률이 있을까용? 뭔가 재사용이 될 것 같아 보이는데 다른 분들 의견도 들어보고 싶네요

wooojini replied:
@Do-ho, @jinhyukoo 해당 부분은 클래스나 모듈로 빼서 재사용하는 부분이 좋아보이네요! 비즈니스 로직에서 state값을 변경할 때 사용될 것 같습니다.

- ex) state.js
```
const ISSUE_STATE = {
    OPEN: "open",
    CLOSED: "closed"
}
const MILESTONE_STATE = {
    OPEN: "open",
    CLOSED: "closed"
}

export { ISSUE_STATE, MILESTONE_STATE  }
```

Do-ho replied:
추후 리팩토링 과정에서 해결하겠습니다.

youngxpepp commented on backend/src/model/label-to-issue.js:
```diff
@@ -0,0 +1,38 @@
+    @PrimaryGeneratedColumn({ type: "int" })
+    labelToIssueId;
+
+    @Column({ type: "int" })
+    @IsInt()
+    issueId;
+
+    @Column({ type: "int" })
+    @IsInt()
+    labelId;
```
밑에 label과 issue 연관관계를 이미 선언하셨군요! 이 부분을 삭제하셔도 괜찮을거 같아요ㅎㅎ
아래에서 FK를 설정하실 수 있습니다😁

youngxpepp commented on backend/src/model/label-to-issue.js:
```diff
@@ -0,0 +1,38 @@
+    @IsDate()
+    deletedAt;
+
+    @ManyToOne(() => Label, (label) => label.labelToIssues)
+    label;
+
+    @ManyToOne(() => Issue, (issue) => issue.labelToIssues)
+    issue;
```
https://typeorm.io/#/relations/joincolumn-options
참조해주세요!

youngxpepp commented on backend/src/model/label-to-issue.js:
```diff
@@ -0,0 +1,38 @@
+
+@Entity()
+class LabelToIssue {
+    @PrimaryGeneratedColumn({ type: "int" })
+    labelToIssueId;
```
다른 곳처럼 increment가 없네요😅 increment 추가해주시면 감사드리겠습니다!

Do-ho replied:
반영했습니다 👍

youngxpepp commented on backend/src/model/milestone.js:
```diff
@@ -0,0 +1,47 @@
+import { Column, CreateDateColumn, Entity, ManyToOne, OneToMany, PrimaryGeneratedColumn, UpdateDateColumn, DeleteDateColumn } from "typeorm";
+import { Issue } from "./issue";
+
+const STATE = {
+    OPEN: "open",
+    CLOSED: "closed"
+};
```
위에서 리뷰 드린 것과 마찬가지에용ㅎㅎ 재사용 가능성이 존재할까요?

youngxpepp commented on backend/src/model/user-to-issue.js:
```diff
@@ -0,0 +1,38 @@
+
+@Entity()
+class UserToIssue {
+    @PrimaryGeneratedColumn({ type: "int" })
+    userToIssueId;
```
여기에도 increment를 넣으면 어떨까요?ㅎㅎ

youngxpepp commented on backend/src/model/user-to-issue.js:
```diff
@@ -0,0 +1,38 @@
+    @PrimaryGeneratedColumn({ type: "int" })
+    userToIssueId;
+
+    @Column({ type: "int" })
+    @IsInt()
+    userId;
+
+    @Column({ type: "int" })
+    @IsInt()
+    issueId;
```
동일 내용 리뷰 드렸습니다!

youngxpepp commented on backend/src/model/user-to-issue.js:
```diff
@@ -0,0 +1,38 @@
+    @IsDate()
+    deletedAt;
+
+    @ManyToOne(() => User, (user) => user.userToIssues)
+    user;
+
+    @ManyToOne(() => Issue, (issue) => issue.userToIssues)
+    issue;
```
동일 내용 리뷰 드렸습니다!

youngxpepp reviewed (changes requested):
정말 고생하셨어요! TypeORM 처음 쓰신거 맞나요?... ㄷㄷ
리뷰 드린 것만 다시 한번 봐주실 수 있을까요😀

wooojini commented:
@Do-ho 방금 테스트 코드를 작성해보니 DB에 실제 insert 되는 순간 결정되는 값들은 모델 객체를 생성하는 당시에는 undefined로 설정되는 것을 확인했습니다! 이런 경우, class-validator로 검증을 하면, 문제가 될 것 같습니다.

```
@Entity({ name: "issue" })
class Issue {
    @PrimaryGeneratedColumn("increment", { type: "int" })
    id;

    @Column({ name: "title", type: "varchar", charset: "utf-8" })
    @IsString()
    title;

    @Column({ name: "content", type: "varchar" })
    @IsString()
    @IsUrl()
    content;

    @Column({ name: "state", type: "varchar", default: STATE.OPEN })
    @IsString()
    state;

    @CreateDateColumn({ name: "created_at", type: "datetime" })
    @IsDate()
    createdAt;

    @UpdateDateColumn({ name: "updated_at", type: "datetime" })
    @IsDate()
    updatedAt;

    @DeleteDateColumn({ name: "deleted_at", type: "datetime" })
    @IsDate()
    deletedAt;
}
export { Issue };
```

 예를 들어 위와 같이 issue 모델이 존재하는데, createdAt, updatedAt, deletedAt처럼 실제 DB에 insert시 자동으로 값을 넣어주는 필드나 state와 같이 default를 설정해서 마찬가지로 insert하는 경우입니다.
 
 문제는  issue모델 객체를 생성하는 당시에는 해당 값이 결정되지 않기 때문에, 유효성 검사를 하는데 error가 발생합니다.

- 아래코드는 issue 테스트코드의 일부입니다.
```
test("유효한 값을 가진 이슈 모델 객체만 새롭게 생성할 수 있다.", async (done) => {
    ....

    const issue = issueRepository.create({ title: "new Issue!", content: "https://test.markdown.io", state: "open" });
    await validate(issue); //Validation Error 발생!

    expect(issue.title).toBe("new Issue!");
    expect(issue.content).toBe("https://test.markdown.io");
    expect(issue.state).toBe("open");
    done();
});
```

- 실제 에러내용 입니다.
```
[
        ValidationError {
          target: Issue {
            id: undefined,
            title: 'new Issue!',
            content: 'https://test.markdown.io',
            state: 'open',
            createdAt: undefined,
            updatedAt: undefined,
            deletedAt: undefined
          },
          value: undefined,
          property: 'createdAt',
          children: [],
          constraints: { isDate: 'createdAt must be a Date instance' }
        },
        ValidationError {
          target: Issue {
            id: undefined,
            title: 'new Issue!',
            content: 'https://test.markdown.io',
            state: 'open',
            createdAt: undefined,
            updatedAt: undefined,
            deletedAt: undefined
          },
          value: undefined,
          property: 'updatedAt',
          children: [],
          constraints: { isDate: 'updatedAt must be a Date instance' }
        },
        ValidationError {
          target: Issue {
            id: undefined,
            title: 'new Issue!',
            content: 'https://test.markdown.io',
            state: 'open',
            createdAt: undefined,
            updatedAt: undefined,
            deletedAt: undefined
          },
          value: undefined,
          property: 'deletedAt',
          children: [],
          constraints: { isDate: 'deletedAt must be a Date instance' }
        }
      ]
```

Do-ho committed:
fix: eager, cascade 등 모델 수정

- 외래키로 설정된 column 재정의 부분 삭제
- join을 사용할 경우 한번의 쿼리 수행으로 연관관계에 있는 엔티티까지 조회해 성능 향상을 위한 eager loading 사용
- 이상 현상을 피하기 위한 cascade 설정

Do-ho committed:
refactor: 공통 타입 상태 변수 선언

- 비즈니즈 로직 상으로 계속적으로 쓰이는 변수를 관리하기 위해 해당 폴더 및 파일 생성

youngxpepp reviewed (approved):
고생하셨습니다! 세부적인 수정사항들은 개발 진행하면서 고쳐나가도 될거 같아요. 감사합니다👍
